src/Symfony/Component/DependencyInjection/Secrets/JweEntry.php

```diff
@@ -0,0 +1,128 @@
+    private function compact()
+    {
+        $valuesForCompaction = array(
+            $this->base64url_encode($this->header),
+            sodium_bin2base64($this->encryptedKey, SODIUM_BASE64_VARIANT_URLSAFE),
+            sodium_bin2base64($this->iv, SODIUM_BASE64_VARIANT_URLSAFE),
+            sodium_bin2base64($this->cipherText, SODIUM_BASE64_VARIANT_URLSAFE),
+            $this->base64url_encode($this->authTag),
+        );
+
+        return implode('.', $valuesForCompaction);
+    }
+
+    private function decryptedSecret(string &$keyPair)
+    {
+        $plaintextCek = sodium_crypto_box_seal_open(
+            $this->encryptedKey,
+            $keyPair
+        );
+
+        $decrypted = sodium_crypto_aead_aes256gcm_decrypt(
+            $this->cipherText,
+            $this->authTag,
+            $this->iv,
+            $plaintextCek
+        );
+        if (false === $decrypted) {
+            throw new RuntimeException(sprintf(
+                'Unable to decrypt secrets. Verify the configured key pair'
+            ));
+        }
+
+        return $decrypted;
+    }
+
+    private function generateHeader()
+    {
+        //TODO: can you force utf8 in string declaration in php?
+        //TODO: figure out correct name for alg
+        return '{"alg":"curve25519xsalsa20poly1305","enc":"A256GCM"}';
+    }
+
+    private function hydrate(string $compactedEntry)
+    {
+        //TODO add validations
+        list($header, $encryptedKey, $iv, $cipherText, $authTag) = explode('.', $compactedEntry);
+        $this->header = $this->base64url_decode($header);
+        $this->encryptedKey = sodium_base642bin($encryptedKey, SODIUM_BASE64_VARIANT_URLSAFE);
+        $this->iv = sodium_base642bin($iv, SODIUM_BASE64_VARIANT_URLSAFE);
+        $this->cipherText = sodium_base642bin($cipherText, SODIUM_BASE64_VARIANT_URLSAFE);
+        $this->authTag = $this->base64url_decode($authTag);
+    }
+
+    private function setEncryptedValues(string $secret, string $pubKey)
+    {
+        $this->header = $this->generateHeader();
+        $rawCek = sodium_crypto_aead_aes256gcm_keygen();
+        $encryptedCek = sodium_crypto_box_seal(
+            $rawCek,
+            $pubKey
+        );
+
+        $this->encryptedKey = $encryptedCek;
+        $nonce = random_bytes(SODIUM_CRYPTO_AEAD_AES256GCM_NPUBBYTES);
+        $this->iv = $nonce;
+        $additionalData = $this->generateHeader();
+        if (sodium_crypto_aead_aes256gcm_is_available()) {
+            $ciphertext = sodium_crypto_aead_aes256gcm_encrypt(
+                $secret,
+                $additionalData,
+                $nonce,
+                $rawCek
+            );
+        } else {
+            $ciphertext = '';
+            //TODO: add alternative encryption method
+        }
+
+        //TODO: figure out if this conversion is needed
+        //mb_convert_encoding($additionalData, "ASCII");
```

Comment:
Is this conversion needed? :)

Reply:
I had originally added this because it was included as a step in the RFC spec but it shouldn't be required here, removing